README.md
# PHP-API-CLIENT
Cliente PHP para consumir el API de Flow.

**Flow.cl** es una pasarela de pagos para comercio electrónico. Este cliente le permite integrar su ecommerce para recibir pagos online.

## Requerimientos
* PHP 5.2 o superior
* php_curl activado
* php_openssl activado

## Instalación
Baje la última versión y copie los archivos en su servidor.

## Documentación
La documentación completa del API REST de Flow la encuentrá aquí: https://www.flow.cl/docs/api.html

## Comenzando
### Configurando el cliente
Configure correctamente el cliente en el archivo **lib/Config.class.php**.
Lo primero que debe configurar es su apiKey y secretKey del comercio registrado en Flow. Esto lo obtiene en la sección **Mis datos** acceda a https://www.flow.cl, una vez autenticado con su cuenta Flow, seleccione Mis datos y recupere su apiKey y secretKey desde la sección Seguridad.
- **APIKEY** el apiKey obtenida desde su cuenta Flow
- **SECRETKEY** el secretKey obtenida desde su cuenta Flow
- **APIURL** la URL del endpoint del API de Flow, Aquí podrá configurar el endpoint de producción o del sandbox. Esta información se obtiene en la documentación del API https://www.flow.cl/docs/api.html
- **BASEURL** La URL base donde instaló el cliente PHP en su servidor


```php
$COMMERCE_CONFIG = array(
 	"APIKEY" => "1F90971E-8276-4713-97FF-2BLF5091EE3B", // Registre aquí su apiKey
 	"SECRETKEY" => "f8b45f9b8bcdb5702dc86a1b894492303741c405", // Registre aquí su secretKey
 	"APIURL" => "https://www.flow.cl/api", // Producción EndPoint o Sandbox EndPoint
 	"BASEURL" => "https://www.misitio/apiFlow" //Registre aquí la URL base en su página donde instaló el cliente
 );
```
### Llamando a un servicio
En este ejemplo crearemos una Orden de Cobro y redireccionaremos el browser del pagador para efectuar el pago
```php
<?php
/**
 * Ejemplo de creación de una orden de cobro, iniciando una transacción de pago
 * Utiliza el método payment/create
 */
require(__DIR__ . "/../../lib/FlowApi.class.php");

//Para datos opcionales campo "optional" prepara un arreglo JSON
$optional = array(
	"rut" => "9999999-9",
	"otroDato" => "otroDato"
);
$optional = json_encode($optional);

//Prepara el arreglo de datos
$params = array(
	"commerceOrder" => rand(1100,2000),
	"subject" => "Pago de prueba",
	"currency" => "CLP",
	"amount" => 5000,
	"email" => "cliente@gmail.com",
	"paymentMethod" => 9,
	"urlConfirmation" => Config::get("BASEURL") . "/examples/payments/confirm.php",
	"urlReturn" => Config::get("BASEURL") ."/examples/payments/result.php",
	"optional" => $optional
);
//Define el metodo a usar
$serviceName = "payment/create";

try {
	// Instancia la clase FlowApi
	$flowApi = new FlowApi;
	// Ejecuta el servicio
	$response = $flowApi->send($serviceName, $params,"POST");
	//Prepara url para redireccionar el browser del pagador
	$redirect = $response["url"] . "?token=" . $response["token"];
	header("location:$redirect");
} catch (Exception $e) {
	echo $e->getCode() . " - " . $e->getMessage();
}
?>
```
Otros ejemplos los podrá ver en la carpeta **examples** de este cliente.


CHANGELOG
af34b4d Tue Oct 29 19:26:53 2019 -0300 Initial commit
1b79039 Fri Oct 11 14:19:47 2019 -0300 Update FlowApi.class.php
36810f9 Fri Jul 26 13:13:43 2019 -0400 Revert "Update FlowApi.class.php"
8738bb6 Fri Jul 26 12:51:17 2019 -0400 Merge branch 'master' of https://github.com/flowcl/PHP-API-CLIENT
ace67f3 Fri Jul 26 12:49:13 2019 -0400 Update FlowApi.class.php
bbbcf0a Thu Jul 26 18:44:06 2018 -0400 Update Read.me
620fe5e Thu Jul 26 15:05:40 2018 -0400 Update Read.me
76dc5b1 Thu Jul 26 15:04:53 2018 -0400 Update Read.me
45debb3 Thu Jul 26 15:02:17 2018 -0400 Actualiza las url
84e45e5 Thu Jul 26 14:49:27 2018 -0400 Merge branch 'master' of https://github.com/flowcl/PHP-API-CLIENT
f9e9414 Thu Jul 26 14:49:21 2018 -0400 Update payments/create
605d20f Thu Jul 26 14:34:55 2018 -0400 Update Read.me
aa46125 Thu Jul 26 13:21:02 2018 -0400 Update Read.me
2a22902 Thu Jul 26 13:14:31 2018 -0400 Update Read.me
da44095 Thu Jul 26 13:08:04 2018 -0400 Agrega información a Read.me
211cd15 Thu Jul 26 12:15:47 2018 -0400 modifica require
f88edbc Thu Jul 26 12:13:39 2018 -0400 Agrega ejemplos payments
5fa37c4 Thu Jul 26 11:08:54 2018 -0400 Agrega componentes lib
0451aa7 Thu Jul 26 09:58:38 2018 -0400 Initial commit


lib/Config.class.php
<?php
/**
 * Clase para Configurar el cliente
 * @Filename: Config.class.php
 * @version: 2.0
 * @Author: flow.cl
 * @Email: csepulveda@tuxpan.com
 * @Date: 28-04-2017 11:32
 * @Last Modified by: Carlos Sepulveda
 * @Last Modified time: 28-04-2017 11:32
 */
 
 $COMMERCE_CONFIG = array(
 	"APIKEY" => "mi apiKey", // Registre aquí su apiKey
 	"SECRETKEY" => "mi secretKey", // Registre aquí su secretKey
 	"APIURL" => "https://www.flow.cl/api", // Producción EndPoint o Sandbox EndPoint
 	"BASEURL" => "https://www.micomercio.cl/apiFlow" //Registre aquí la URL base en su página donde instalará el cliente
 );
 
 class Config {
 	
	static function get($name) {
		global $COMMERCE_CONFIG;
		if(!isset($COMMERCE_CONFIG[$name])) {
			throw new Exception("The configuration element does not exist", 1);
		}
		return $COMMERCE_CONFIG[$name];
	}
 }


lib/FlowApi.class.php
<?php

/**
 * Clase cliente del Api2 de Flow
 * @Filename: FlowApi.class.php
 * @version: 2.1
 * @Author: flow.cl
 * @Email: csepulveda@tuxpan.com
 * @Date: 28-04-2017 11:32
 * @Last Modified by: Carlos Sepulveda
 * @Last Modified time: 28-04-2017 11:32
 */

require(__DIR__ . "/Config.class.php");

class FlowApi {
	
	protected $apiKey;
	protected $secretKey;
	
	
	public function __construct() {
		$this->apiKey = Config::get("APIKEY");
		$this->secretKey = Config::get("SECRETKEY");
	}
	
	
	/**
	 * Funcion que invoca un servicio del Api de Flow
	 * @param string $service Nombre del servicio a ser invocado
	 * @param array $params datos a ser enviados
	 * @param string $method metodo http a utilizar
	 * @return string en formato JSON
	 */
	public function send( $service, $params, $method = "GET") {
		$method = strtoupper($method);
		$url = Config::get("APIURL") . "/" . $service;
		$params = array("apiKey" => $this->apiKey) + $params;
		$params["s"] = $this->sign($params);
		if($method == "GET") {
			$response = $this->httpGet($url, $params);
		} else {
			$response = $this->httpPost($url, $params);
		}
		
		if(isset($response["info"])) {
			$code = $response["info"]["http_code"];
			if (!in_array($code, array("200", "400", "401"))) {
				throw new Exception("An unexpected HTTP_CODE error occurred: " .$code , $code);
			}
		}
		$body = json_decode($response["output"], true);
		return $body;
	}
	
	/**
	 * Funcion para setear el apiKey y secretKey y no usar los de la configuracion
	 */
	public function setKeys($apiKey, $secretKey) {
		$this->apiKey = $apiKey;
		$this->secretKey = $secretKey;
	}
	
	
	/**
	 * Funcion que firma los parametros
	 * @param string $params Parametros a firmar
	 * @return string de firma
	 */
	private function sign($params) {
		$keys = array_keys($params);
		sort($keys);
		$toSign = "";
		foreach ($keys as $key) {
			$toSign .= $key . $params[$key];
		}
		if(!function_exists("hash_hmac")) {
			throw new Exception("hash_hmac function does not exist", 1);
		}
		return hash_hmac('sha256', $toSign , $this->secretKey);
	}
	
	
	/**
	 * Funcion que hace el llamado via http GET
	 * @param string $url url a invocar
	 * @param array $data datos a enviar
	 * @param string $sign firma de los datos
	 * @return string en formato JSON 
	 */
	private function httpGet($url, $params) {
		$url = $url . "?" . http_build_query($params);
		$ch = curl_init();
		curl_setopt($ch, CURLOPT_URL, $url);
		curl_setopt($ch, CURLOPT_RETURNTRANSFER, TRUE);
		$output = curl_exec($ch);
		if($output === false) {
			$error = curl_error($ch);
			throw new Exception($error, 1);
		}
		$info = curl_getinfo($ch);
		curl_close($ch);
		return array("output" =>$output, "info" => $info);
	}
	
	/**
	 * Funcion que hace el llamado via http POST
	 * @param string $url url a invocar
	 * @param array $data datos a enviar
	 * @param string $sign firma de los datos
	 * @return string en formato JSON 
	 */
	private function httpPost($url, $params ) {
		$ch = curl_init();
		curl_setopt($ch, CURLOPT_URL, $url);
		curl_setopt($ch, CURLOPT_RETURNTRANSFER, TRUE);
		curl_setopt($ch, CURLOPT_POST, TRUE);
		curl_setopt($ch, CURLOPT_POSTFIELDS, $params);
		$output = curl_exec($ch);
		if($output === false) {
			$error = curl_error($ch);
			throw new Exception($error, 1);
		}
		$info = curl_getinfo($ch);
		curl_close($ch);
		return array("output" =>$output, "info" => $info);
	}
	
	
}


examples/payments/createEmail.php
<?php
/**
 * Ejemplo de creación de una orden de cobro vái cobro por email
 * Utiliza el método payment/createEmail
 */
require(__DIR__ . "/../../lib/FlowApi.class.php");

//Prepara los parámetros
$params = array(
	"commerceOrder" => rand(1100,2000),
	"subject" => "pago prueba cobro Email",
	"currency" => "CLP",
	"amount" => 2000,
	"email" => "cliente@gmail.com",
	"paymentMethod" => 9,
	"urlConfirmation" => Config::get("BASEURL") . "/examples/payments/confirm.php",
	"urlReturn" => Config::get("BASEURL") ."/examples/payments/result.php",
	"forward_days_after" => 1,
	"forward_times" => 2
);

$serviceName = "payment/createEmail";

try {
	$flowApi = new FlowApi;
	
	$response = $flowApi->send($serviceName, $params,"POST");
	
	print_r($response);
	
	
} catch (Exception $e) {
	echo $e->getCode() . " - " . $e->getMessage();
}

?>

examples/payments/create.php
<?php
/**
 * Ejemplo de creación de una orden de cobro, iniciando una transacción de pago
 * Utiliza el método payment/create
 */
require(__DIR__ . "/../../lib/FlowApi.class.php");

//Para datos opcionales campo "optional" prepara un arreglo JSON
$optional = array(
	"rut" => "9999999-9",
	"otroDato" => "otroDato"
);
$optional = json_encode($optional);

//Prepara el arreglo de datos
$params = array(
	"commerceOrder" => rand(1100,2000),
	"subject" => "Pago de prueba",
	"currency" => "CLP",
	"amount" => 5000,
	"email" => "cliente@gmail.com",
	"paymentMethod" => 9,
	"urlConfirmation" => Config::get("BASEURL") . "/examples/payments/confirm.php",
	"urlReturn" => Config::get("BASEURL") ."/examples/payments/result.php",
	"optional" => $optional
);
//Define el metodo a usar
$serviceName = "payment/create";

try {
	// Instancia la clase FlowApi
	$flowApi = new FlowApi;
	// Ejecuta el servicio
	$response = $flowApi->send($serviceName, $params,"POST");
	//Prepara url para redireccionar el browser del pagador
	$redirect = $response["url"] . "?token=" . $response["token"];
	header("location:$redirect");
} catch (Exception $e) {
	echo $e->getCode() . " - " . $e->getMessage();
}

?>